controllers/workflow_controller_test.go

```diff
@@ -0,0 +1,195 @@
+		wf = &dwsv1alpha1.Workflow{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      wfid,
+				Namespace: corev1.NamespaceDefault,
+			},
+			Spec: dwsv1alpha1.WorkflowSpec{
+				DesiredState: dwsv1alpha1.StateProposal,
+				WLMID:        "test",
+				JobID:        0,
+				UserID:       0,
+				GroupID:      0,
+				DWDirectives: []string{},
+			},
+			Status: dwsv1alpha1.WorkflowStatus{
+				State:  dwsv1alpha1.StateProposal,
+				Ready:  false,
+				Status: dwsv1alpha1.StatusDriverWait,
+			},
+		}
+	})
+
+	AfterEach(func() {
+		if wf != nil {
+			Expect(k8sClient.Delete(context.TODO(), wf)).To(Succeed())
+
+			wfExpected := &dwsv1alpha1.Workflow{}
+			Eventually(func() error { // Delete can still return the cached object. Wait until the object is no longer present.
+				return k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(wf), wfExpected)
+			}).ShouldNot(Succeed())
+		}
+	})
+
+	It("Can complete Workflow driver states", func() {
+		action := "complete"
+		wf.Spec.DWDirectives = []string{
+			fmt.Sprintf("#DW STATUS action=%s", action),
+		}
+
+		driverStatus := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+		}
+
+		// Set a bunch of other statuses that should remain unchanged
+		statusWrongDriver := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   "notMyDriver",
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+		}
+		statusWrongWatchState := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateDataIn,
+		}
+		statusCompleted := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+			Completed:  true,
+		}
+		statusError := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+			Status:     dwsv1alpha1.StatusError,
+		}
+		wf.Status.Drivers = []dwsv1alpha1.WorkflowDriverStatus{
+			driverStatus,
+			statusWrongDriver,
+			statusWrongWatchState,
+			statusCompleted,
+			statusError,
+		}
+
+		ct := metav1.NowMicro()
+		expectedDriverStatuses := []dwsv1alpha1.WorkflowDriverStatus{
+			{
+				DriverID:     DRIVERID,
+				DWDIndex:     0,
+				WatchState:   dwsv1alpha1.StateProposal,
+				Status:       dwsv1alpha1.StatusCompleted,
+				Completed:    true,
+				CompleteTime: &ct,
+			},
+			statusWrongDriver,
+			statusWrongWatchState,
+			statusCompleted,
+			statusError,
+		}
+
+		Expect(k8sClient.Create(context.TODO(), wf)).To(Succeed())
+
+		Eventually(func(g Gomega) []dwsv1alpha1.WorkflowDriverStatus {
+			returnWf := &dwsv1alpha1.Workflow{}
+			g.Expect(k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(wf), returnWf)).To(Succeed())
+			driverStatus = returnWf.Status.Drivers[0]
+			return returnWf.Status.Drivers
+		}).Should(BeComparableTo(expectedDriverStatuses,
+			cmp.Comparer(func(x, y *metav1.MicroTime) bool {
```

Comment:
Wouldn't you just want to see that the status is completed?

Reply:
The test driver is also setting the completed time.